ckanext/dia/commands.py

```diff
@@ -42,17 +42,19 @@ def help(self):
         print self.__doc__
 
     def cleanup_datastore(self):
-        # E.B 15/3/18 HACK: running the datastore 20 times in a row allows us to 
-        # get datastore table cleanup to work. Without this hack only 300 tables 
+        # E.B 15/3/18 HACK: running the datastore 20 times in a row allows us to
+        # get datastore table cleanup to work. Without this hack only 300 tables
         # will be cleaned up.
+        start_offset = 0
         for i in xrange(20):
             print 'invoking iteration %s of the cleanup_datastore_once function' % i
-            deletes, errors = self.cleanup_datastore_once()
+            deletes, errors, total_tested = self.cleanup_datastore_once(start_offset)
+            start_offset += total_tested - errors - deletes
```

Comment:
`cleanup_datastore_once` now returns the total number of resource records it tested during it's run. Given that the order of the data returned from the query to get all the resources the datastore references seems to be ordered deterministically, we can start the next round of cleaning at a point that we know represents the last file that was tested.
e.g. [1,2,3,4,5] we test the first three and delete 1 and 3. The next time the scan runs we have a new set [2,4,5] and we start at index 3 - 0 - 2 = 1. This means we only test 4 and 5 in the second run, rather than testing 2 again. This is a huge time saver when we are talking about thousands of records and the api calls to check for existence `resource_show` take the longest when the item does exist.